File: identity-verification/js/src/models/identity-record.ts
import {PublicKey} from "@solana/web3.js";
import {Status} from "./status";

export type BumpSeed = number

export class IdentityRecord {

    // ============================================================
    // === Public Static API ======================================
    // ============================================================

    /**
     * Factory method that takes json and produces an identity record instance
     * @param info
     */
    static with(info: any): IdentityRecord {

        return new IdentityRecord(
            info.bump,
            info.iaStatus,
            info.amlStatus,
            info.kycStatus,
            info.authority
        )

    }

    // ============================================================
    // === Public API =============================================
    // ============================================================

    // Public Properties

    /// bump seed used in deriving the pda for the status account
    bump: BumpSeed

    // Accreditation status of the user
    iaStatus: Status

    /// AML status of the user
    amlStatus: Status

    /// KYC status of the user
    kycStatus: Status

    /// Account who has update authority over the account
    authority: PublicKey

    // ============================================================
    // === Private API ============================================
    // ============================================================

    // Private Methods

    /**
     * @param bump
     * @param iaStatus
     * @param amlStatus
     * @param kycStatus
     * @param authority
     */
    private constructor(
        bump: BumpSeed,
        iaStatus:Status,
        amlStatus: Status,
        kycStatus: Status,
        authority: PublicKey
    ) {
        this.bump = bump
        this.iaStatus = iaStatus
        this.amlStatus = amlStatus
        this.kycStatus = kycStatus
        this.authority = authority
    }

}

File: identity-verification/program/src/state/identity_record.rs
use anchor_lang::prelude::*;

#[derive(Default)]
#[account]
pub struct IdentityRecord {
    /// bump seed used in deriving the pda for the status account
    pub bump: u8,

    /// Accreditation status of the user (0 == initial, 1 == started, 2 == approved, 3 == rejected).
    pub ia_status: u8,

    /// AML status of the user (0 == initial, 1 == started, 2 == approved, 3 == rejected).
    pub aml_status: u8,

    /// KYC status of the user (0 == initial, 1 == started, 2 == approved, 3 == rejected).
    pub kyc_status: u8,

    /// Account who has update authority over the account
    pub authority: Pubkey,
}

impl IdentityRecord {
    pub const LEN: usize = 8 + // discriminator
        1 + // bump
        (1 * 3) + // statuses
        32; // authority pubkey
}
